app/src/main/java/de/xikolo/storages/RecentCoursesStorage.kt

```diff
@@ -0,0 +1,56 @@
+package de.xikolo.storages
+
+import android.content.Context
+import androidx.preference.PreferenceManager
+import com.google.gson.Gson
+import com.google.gson.reflect.TypeToken
+import de.xikolo.App
+import de.xikolo.storages.base.BaseStorage
+import de.xikolo.storages.base.SharedPreferenceLiveData
+
+typealias RecentCourse = Pair<String, String>
+
+const val MAX_SHORTCUTS = 3
+
+class RecentCoursesStorage : BaseStorage(PREF_RECENT_COURSES, Context.MODE_PRIVATE) {
+
+    val recentCourses: LinkedHashSet<RecentCourse>
+        get() {
+            val json = getString(PREF_RECENT_COURSES)
+            val type = object : TypeToken<LinkedHashSet<RecentCourse>>() {}.type
+            if (json == null || type == null) {
+                return LinkedHashSet<RecentCourse>()
+            }
+            println(json)
+            return Gson().fromJson(json, type)
+        }
+
+    val coursesLive: SharedPreferenceLiveData<String> by lazy {
+        // what name does the preference have? What is the sharedPref Parameter supposed to do?
+        SharedPreferenceLiveData.SharedPreferenceStringLiveData(PreferenceManager.getDefaultSharedPreferences(App.instance), PREF_RECENT_COURSES, "")
+    }
+
+    fun addCourse(courseId: String, title: String) {
+        val courses: LinkedHashSet<RecentCourse> = recentCourses
+
+        when {
+            courses.contains(Pair(courseId, title)) -> {
+                courses.remove(Pair(courseId, title))
+                courses.add(Pair(courseId, title))
+            }
+            courses.size < MAX_SHORTCUTS -> {
+                courses.add(Pair(courseId, title))
+            }
+            courses.size == MAX_SHORTCUTS -> {
+                courses.remove(courses.last())
+                courses.add(Pair(courseId, title))
+            }
+        }
```

Comment:
Am I right assuming the newest element is at the last index?
Then following code should do it as well:
```suggestion
        courses = courses.also { it.add(Pair(courseId, title)) }.takeLast(MAX_SHORTCUTS)
```

Reply:
Your assumption is right, but your code allows the same course to be added multiple times, isn't it? In the case that I visit a course 3 times in a row for example. But if it checks first, if the course is already in the list, your code would make a shorter else branch. 

Something like:
```
if (courses.contains(Pair(courseId, title))) {
                courses.remove(Pair(courseId, title))
                courses.add(Pair(courseId, title))
 } else {
                courses = courses.also { it.add(Pair(courseId, title)) }.takeLast(MAX_SHORTCUTS)
}
```

Or is there also a shortcut possible using scope functions? So maybe:
```
courses = courses.also { 
   if (courses.contains(Pair(courseId, title))) {
                courses.remove(Pair(courseId, title))
   } 
   it.add(Pair(courseId, title)) }.takeLast(MAX_SHORTCUTS)
}
```

Reply:
Try something like:
```kotlin
courses = courses
    .filter { it.first != courseId }
    .also { it.add(Pair(courseId, title)) }
    .reversed()
    .take(MAX_SHORTCUTS)
    .reversed()
```